core/consensus/grandpa/vote_graph/vote_graph_impl.cpp

```diff
@@ -243,35 +243,42 @@ namespace kagome::consensus::grandpa {
       return boost::none;
     }
 
-    // breadth-first search starting from this node.
-    bool loop = true;
-    while (loop) {
-      using namespace boost::adaptors;  // NOLINT
-      auto filtered_entries =
-          active_node.descendents | transformed([this](const BlockHash &d) {
-            return std::make_pair(d, entries_.at(d));
-          })
-          | filtered([&](const auto &pair) {
-              const auto &[_, node] = pair;
-              if (force_constrain && current_best) {
-                return inDirectAncestry(
-                    node, current_best->block_hash, current_best->block_number);
-              }
-              return true;
-            });
-      const auto &next_descendent_it = boost::range::find_if(
-          filtered_entries, [&condition](const auto &pair) {
-            const auto &[_, node] = pair;
-            return condition(node.cumulative_vote);
-          });
-      if (next_descendent_it == filtered_entries.end()) {
-        break;
-      }
-      force_constrain = false;
-      const auto &[key, node] = *next_descendent_it;
-      node_key = key;
-      active_node = node;
-    }
+    std::function<void(const VoteGraph::Entry &entry)> search =
+        [&](const VoteGraph::Entry &entry) {
```

Comment:
You don't need recursive call for BFS, it can be well-implemented with FIFO-queue. Recursive calls can not be inlined and well-optimized. So it will lead to stack frame creation for each node. More over std::function is slow on its own. I guess that this code must be performance-critical?

Reply:
Hm... You are right, there is good mechanic to make it without recursion.
Thanks, I'll do it.